torch_xla/experimental/gpu.py

```diff
@@ -0,0 +1,43 @@
+import os
+import torch_xla
+import torch_xla.core.xla_env_vars as xenv
+
+distributed_service = None
+
+
+def num_local_processes() -> int:
+  """Returns number of processes to create on this host.
+    
+  Raises:
+    AssertionError: if GPU_NUM_DEVICES environment variable
+                    is not configured
+  """
+  assert xenv.GPU_NUM_DEVICES in os.environ, \
+      "Must set `GPU_NUM_DEVICES` environment variable to use the PjRt GPU client"
+  return int(os.environ[xenv.GPU_NUM_DEVICES])
+
+
+def initialize_distributed_runtime(global_world_size: int) -> None:
+  """Configures GPU distributed runtime parameters.
+
+  Must be run before using any XLA devices.
+
+  Args:
+    global_world_size: number of devices in the cluster.
+  """
+  if global_world_size > 1:
+    # TODO(jonbolin): For multi-host, this needs to be consistent across hosts
+    os.environ.setdefault(xenv.PJRT_DIST_SERVICE_ADDR, '127.0.0.1:8547')
+    global distributed_service
+    if distributed_service is None:
+      num_nodes = global_world_size
+      distributed_service = torch_xla._XLAC._xla_get_distributed_runtime_service(
+          num_nodes)
+
+
+def shutdown_distributed_runtime() -> None:
```

Comment:
What happens if we crash before shutting down the runtime?

Reply:
By default, the runtime is shut down in the destructor, but since we need to recreate the service for each `xmp.spawn` call, I decided to make an explicit `shutdown` call here instead of relying on the garbage collector to avoid potential race conditions.

In a crash of the main Python process (which hosts the DistributedRuntimeService), the individual per-device subprocesses should be terminated as children.

If one of the subprocesses crashes, the server will become aware of the crash through missed heartbeats. This is actually what happens with the current collectives issue, since the subprocesses crash after the replica groups are changed. In that case, I've observed the Python process will hang for a bit before terminating with a BrokenProcessPool exception in Python.

Reply:
Thanks for explanation Jon!